Fixes for PR #458 (Multisig: Connect request and SignMultisigTransaction request)

danimoh opened the pull request:
Addresses most of the comments on PR #458.

danimoh committed:
Remove references to non-existent src/lib/multisig/wasm/pkg

That file was removed in fbbcf06.

danimoh committed:
PublicRequest,Keyguard.d.ts: small typing improvements

danimoh committed:
Remove forge.min.js.map as it contains no actually useful mappings

danimoh committed:
Key: security fix for RSA key seed generation

Previoulsy, the bytes have been interpreted as UTF8 bytes and encoded to a Javascript string,
which is UTF16 encoded. Up to 4 UTF8 bytes can represent a single unicode symbol, which is then
encoded as a single UTF16 character or a UTF16 surrogate pair. The following re-encodings are
possible:
1 UTF8 byte -> 1 UTF16 character
2 UTF8 bytes -> 1 UTF16 character
3 UTF8 bytes -> 1 UTF16 character
4 UTF8 bytes -> 2 UTF16 characters (surrogate pair)

This means that the resulting string could be as short as a third of the length of the input byte
array. As node-forge expects a binary string as input, where each character is interpreted as a
single byte, overall in the worst case, two thirds of the original data were discarded and an
exponentially weaker entropy was used.

This fix is a breaking change, as the resulting RSA key will be different. Therefore, users who
already have a stored RSA key need to logout to clear their old RSA key (or we decide to rename the
property in the store). Any app that already got the old RSA public key via a Connect request must
discard its known RSA public keys and associated encrypted data, and connect the account again to
get the new RSA public key.

danimoh committed:
Keys,RSAKeysIframe: use Uint8Buffers to communicate data

Binary strings are not very intuitive (and have been used incorrectly) and are not what
our APIs usually use.
Also now instead of copying the data, transfer ownership on postMessage communications.
Additionally, more thoroughly check the communicated data to be of the expected format.

danimoh committed:
RSAKeysIframe: add a content security policy

danimoh committed:
RSAKeysIframe: refuse running if not in a sandboxed iframe

Otherwise, a malicious page can open the page as iframe (unless disallowed by
an appropriate CSP) or as a popup (always possible, I think), and communicate
with it.

danimoh committed:
Key,RSAKeysIframe: security fix for postMessage communication with RSA iframe

Reject any messages in communication between the RSA iframe and main window, which are not
from / to the intended sender / recipient. Otherwise, the following attack was possible:
A malicious site starts a Connect request in an iframe (via Hub, which then redirects to the
Keyguard). Then, when the malicious site suspects that the Keyguard is waiting for the RSA key
calculation, it itself sends a postmessage to the popup with an RSA key known to the attacker,
which the Keyguard would willingly store and continue to use, also for other websites.

danimoh committed:
RSAKeysIframe: refuse key generation if not enough entropy was provided

The caller should make sure to provide enough entropy or stretch it out with a proper
key derivation function, instead of us simply repeating the seed.

Additionally, don't truncate entropy if more was provided than necessary, as it is
been made use of nonetheless.

danimoh committed:
RSAKeysIframe: add notes on why RSA generation must be single threaded

danimoh committed:
Key: some code cleanup

- Rename _isEncryptionKeyParamsEqual to _areEncryptionKeyParamsEqual and
  _isDefaultEncryptionKeyParams to _areDefaultEncryptionKeyParams and mark
  them as private.
- Make _rsaKey a private property. For external access add a method
  getRsaKeypairIfExists.
- Make collection of otherPublicKeys and otherCommitments in signPartially
  more concise.
- Deduplicate code for setting the RSA key into method setRsaKeyPair.
- Make usage of promises in _computeRsaKeyPair nicer.

danimoh committed:
Key: add documentation why we do RSA computations in a sandboxed iframe

danimoh committed:
Key: remove the RSA iframe after key generation

danimoh committed:
I18n: support translation variables / placeholders

danimoh committed:
I18n: support translating to dom, including placeholders and auto lang updates

danimoh committed:
Obversable: fix handling of deleted listeners, e.g. for I18n observer

danimoh committed:
Connect: i18n fixes and improvements

- Add i18n support to some texts that were missing it.
- Translate full headline instead of individual fragments. This provides more
  context to translators and supports languages where the structure of the
  sentence needs to change.

danimoh committed:
Connect: some code cleanup

- Remove an unused button.
- Preserve stack trace on CoreError.
- Access key params without a necessary type assertion.
- Return meta data on the key algorithm directly from the key, instead of
  hardcoding it.
- Capitalization of error messages

danimoh committed:
Connect: provide more accurate information on exposed addresses

Previously, the UI wrongly claimed, that always the first address is exposed, which
is only the most common case. In fact though, an app can request arbitrary addresses
and an arbitrary number of addresses. In such case, the UI now shows the number of
requested addresses and their derivation paths.

danimoh committed:
Connect: security fix for blind signing of arbitrary messages

The Connect request blindly signed any arbitrary challenge as a valid Nimiq message
signature. I.e. an attacker could trick the user into unknowingly signing messages
with the only restriction enforced by request parsing being that the message must be
textual, which is the most common case for messages to be approved by the user. This
could be abused for impersonating the user and for example authenticating at staking
pool websites as the user.

To avoid this, a distinct prefix separate to the Nimiq message signature prefix is
now used.

It's a breaking change, which also requires adaptions to the signature verification
in apps using the Connect request.

This commit also improves the way message prefixes are shared between the Keyguard
client and the main Keyguard source code by redefining the message prefixes in a
type-checked fashion in the main code, instead of the client having to import it as
a javascript file and exposing the value in a hacky global variable on window.

danimoh committed:
Connect,KeyguardCommand: replicate KeyguardCommand as enum in js

Instead of working only with string values and a separate .d.ts file exporting
KeyguardCommand in an additional namespace, and only type asserting, but not type
checking the strings, replicate KeyguardCommand as a fully type-checked enum on
the js side.

danimoh committed:
ClientEnums: replicate client enums in a more concise fashion

Instead of type checking each enum value individually, typecheck the entire enum
definition at once, and define a utility function for enum replication for nicer
reading code.

danimoh committed:
Connect: add staking and polygon requests to requestable permissions

danimoh committed:
Connect: disable request permissions for now

The were only part of the API but not processed in a meaningful fashion:
They were not shown in the UI for confimation by the user, and were not persisted,
and requests did not check for granted permissions.

danimoh committed:
SignMultisigTransactionApi: stricter and improved request parsing

- Remove redundant null checks.
- Use existing type MultisigConfig['signers'] instead of redefining it.
- For signers, change an InvalidRequestError to a regular Error, as it will be
  mapped to an InvalidRequestError on catch anyways.
- For secrets, ensure that any thrown error is mapped to an InvalidRequestError,
  especially from the RandomSecret constructor.
- More thoroughly check structure of secrets object for encrypted secrets: check
  that it's an object, non null, has the expected properties, and that keyParams
  is an object.
- Use Array.every instead of Array.some to not have to negate the condition.
- Make multisig address validation more concise and more readable with added
  comments.
- Avoid a @ts-ignore in layout parsing.

danimoh committed:
SignMultisigTransaction: remove unused dependencies CurrencyInfo and FiatApi

danimoh committed:
SignMultisigTransaction: some code cleanup

- Remove an unnecessary eslint-disable rule.
- Remove a duplicate newline.
- Remove unnecessary clearing of login file icon's HTML content.
- Preserve stack trace on CoreError.
- Deduplicate loop for collecting own commitment pairs.
- Read key algorithm from key, instead of hardcoding it.
- Simplify filtering other signers.

danimoh committed:
SignMultisigTransaction: i18n improvements

Translate name section as full sentence instead of individual fragments.
This provides more context to translators and supports languages where the
structure of the sentence needs to change.

danimoh committed:
Connect,SignMultisigTransaction: reduce repetition in loading spinner SVG

Apply the same little optimization from 3b69993 to the Connect and
SignMultisigTransaction requests.

danimoh committed:
SignMultisigTransaction: deduplicate style between SignTransaction and SignMultisigTransaction

danimoh committed:
SignMultisigTransaction: show multisig badge on correct side of the transaction

If the multisig is the recipient of the transaction, for example for a remove-stake
transaction, show the multisig badge on the recipient side.

danimoh committed:
SignTransaction,SignMultiSigTransaction: support decoding advanced tx data

- Don't hide information on sender type, recipient type or sender data from the user.
  Instead, try to represent them in a user-friendly fashion.
- Decode transaction data like contract creation data or staking data, instead of
  displaying them only as hex.
- Allow decoding data as utf8 only for regular transactions.
- Provide info on contract withdrawals.
- Deduplicate code for transaction data formatting.
- Update formatted data on language change.

danimoh committed:
SignTransactionApi: for staking tx disallow the SignTransaction request

The SignStaking request should be used instead, which provides a specialized UI for it.

danimoh committed:
AddressInfo: i18n for multisig badge

danimoh committed:
AddressInfo.css: remove invalid css comment and hardcoded value

danimoh committed:
LoginFileIcon: make the address mandatory in the constructor

The class was not used anywhere without providing the address in the constructor.

danimoh committed:
build.sh: add integrity checks for RSAKeysIframe

danimoh committed:
PublicRequest: add documentation on MultisigConfig

danimoh committed:
PublicRequest,Keyguard.d.ts,Keys: rename EncryptionKeyParams to RsaKeyParams

- Avoid confusion with other encryption algorithms used in KeyStore.
- RSA is the only encryption algorithm supported by Connect and SignMultisigTransaction.
  Although the APIs / types are partially designed for supporting different algorithms, they
  aren't really, not entirely and not consequently designed for it, and the name change
  also doesn't change anything about the possibility to support additional algorithms in
  the future.
- More consistent naming with other RSA related methods.
- Rename _areEncryptionKeyParamsEqual to _areEqualRsaKeyParams and _areDefaultEncryptionKeyParams
  to _areDefaultRsaKeyParams.

danimoh committed:
KeyStore: store RSA key encrypted

danimoh committed:
SignMultisigTransaction: check that number of commitments and secrets match

danimoh committed:
KeyStore,Key: use AES for RSA key encryption

danimoh committed:
Key,Keystore: represent RSA key pair as actual CryptoKeys

danimoh committed:
Key: split and rename functionality of setRsaKeyPair

Improve code readability and use more descriptive function names.

sisou commented on src/request/sign-multisig-transaction/SignMultisigTransaction.js:
```diff
@@ -229,50 +224,44 @@ class SignMultisigTransaction {
             request.transaction.serializeContent(),
             ownCommitmentPairs,
-            request.multisigConfig.signers.filter(signer => !signer.publicKey.equals(publicKey)),
+            request.multisigConfig.signers.filter(signer => signer !== ownSigner),
```
This may be a problem, because `Nimiq.KeyPair`s cannot be compared as objects, as they are references into Rust WASM memory. So the same KeyPair can be represented as different objects, with different pointers, altough they represent the same bytes.

I think the previous comparison was more accurate.

danimoh replied:
The `ownSigner` is assigned one of the objects from `request.multisigConfig.signers`, i.e. the strict equality check by reference works as expected.

sisou replied:
Yeah, I noticed that later and changed it back to your version, just in a different place and with an explainer in my latest PR. Sorry, forgot to comment here.

sisou commented on src/lib/TransactionDataFormatting.js:
```diff
@@ -0,0 +1,122 @@
+            return I18n.translatePhrase('tx-data-contract-withdrawal', {
+                contractType: TransactionDataFormatting._capitalize(plainTransaction.senderType),
+            });
+        }
```
Vesting withdrawal transactions can have a UTF8 message as the only data. How is that represented by this? I think only as hex, which is not useful.

danimoh replied:
Yes, hex, I believe. Decoding the Utf8 message makes sense then for this case :+1:.
It's not only vesting withdrawals though, HTLC withdrawals can have additional data as well, apart from the required redeeming information, which is in the signature proof. For `RemoveStake` and `DeleteValidator` transactions, I think additional recipient data might also be allowed, apart from the required sender data, but I have not confirmed.

Your proposed changes handle this correctly though.

sisou commented on src/lib/TransactionDataFormatting.js:
```diff
@@ -0,0 +1,122 @@
+     * @returns {string}
+     */
+    static _formatPlainData(plainData, includeDataType) {
+        if (plainData.type === 'raw') return plainData.raw; // Return plain hex. Empty if transaction has no data.
```
I think this should attempt to decode as UTF8 and only return the raw HEX when decoding fails.

sisou replied:
With my changes to the `formatTransactionData` function in my branch this is now no longer necessary, as this method is only called when the recipient is a contract or the data is not UTF8 anyway, so UTF8 decoding is not relevant.

danimoh replied:
Yes, I like interpreting data as UTF8 if the recipient is a basic account. I'm not a fan of trying to interpret other data as UTF8. So I like your latest approach.

sisou commented on src/lib/TransactionDataFormatting.js:
```diff
@@ -0,0 +1,122 @@
+        } else if (transaction.recipientType === Nimiq.AccountType.Staking) {
+            if (!(transaction.flags & Nimiq.TransactionFlag.Signaling)) { // eslint-disable-line no-bitwise
+                // A staking transaction which creates/adds actual stake by moving funds to the staking contract
+                prefix = I18n.translatePhrase('tx-data-contract-creation', { contractType: 'Staking' });
```
This will translate to "Staking creation", even when users just _add_ stake to their existing stake. That is misleading.

I think the staking contract should have a separate translation, because the staking contract is never _created_, it already exists since the PoS genesis block.

danimoh replied:
Yeah, I was just re-using the translation, but a separate, more specific translation might make sense.

sisou commented on src/request/sign-transaction/SignTransactionApi.js:
```diff
@@ -57,6 +58,11 @@ class SignTransactionApi extends TopLevelApi {
 
+        if (parsedRequest.transaction.senderType === Nimiq.AccountType.Staking
+            || parsedRequest.transaction.recipientType === Nimiq.AccountType.Staking) {
+            throw new Errors.InvalidRequestError('For staking transactions, use a Keyguard sign-staking request');
```
Reword: `'For staking transactions, use the Keyguard request "sign-staking"'`

sisou commented on src/lib/rsa/sandboxed/RSAKeysIframe.html:
```diff
@@ -3,7 +3,8 @@
     <!-- Note: sandbox and frame-ancestors policies can not actually be specified via a meta tag. You have to send them
-    as response headers. -->
+    as response headers. Additionally, the server needs to include the sandboxed iframe's null origin in its cors header
+    for the script files imported here, in order to pass integrity checks added by the build process. -->
```
What does that mean? What must the server send now? Can you list that here so we can tell IT?

danimoh replied:
- [`sandbox` and `frame-ancestors` headers](https://github.com/nimiq/keyguard/pull/458#discussion_r1939751795)
- The `forge.min.js` and `RSAKeysIframe.js` (or rather the deployed versions with filename hashes) need to be send with `Access-Control-Allow-Origin: null`. Mdn has a [warning regarding doing this](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Access-Control-Allow-Origin#null), but as far as I can see, there is no actual damage in doing so in this case.

CC @curdbecker

sisou commented on tools/build.sh:
```diff
@@ -360,13 +360,31 @@ awk '
+# Build RSA iframe
+output "🔑  Building RSA Iframe"
+cp src/lib/rsa/sandboxed/forge.min.js dist/lib/rsa/sandboxed/forge.min.HASH.js
+# Integirty check that forge.min.js is from node-forge 1.3.1
```
Spelling: "Integrity"

sisou commented on tools/build.sh:
```diff
@@ -360,13 +360,31 @@ awk '
+output "🔑  Building RSA Iframe"
+cp src/lib/rsa/sandboxed/forge.min.js dist/lib/rsa/sandboxed/forge.min.HASH.js
+# Integirty check that forge.min.js is from node-forge 1.3.1
+RSA_IFRAME_FORGE_HASH="3Gf9EyQnrZbJZmyESzlWVBPEDdsfLQY8U1Evv204ff0="
```
Why hard-code the hash here but not verify it? That just means it will fail at runtime if we update the script, instead of at build time. 

Either:
1. Calculate the hash from the script that is copied to the build folder
2. Verify this hash against the file it is supposed to represent

danimoh replied:
- Checking it at runtime is the important check for detecting files tampered with, checking it also at build time adds little to that.
- Having a hard coded hash in the code means that it needs to be updated manually in the code, which will show up in pull requests on changes, and serves as a nice reminder on code review to cross-check that file hash with the `forge.min.js` obtained from the original source. If the hash is calculated automatically, a reviewer has to not forget himself to cross check the file `forge.min.js` against the original source, if it's changed in the keyguard code base.

sisou commented on tools/build.sh:
```diff
@@ -360,13 +360,31 @@ awk '
+RSA_IFRAME_SCRIPT_NAME=$(add_hash_to_file_name dist/lib/rsa/sandboxed/RSAKeysIframe.HASH.js)
+cp src/lib/rsa/sandboxed/RSAKeysIframe.html dist/lib/rsa/sandboxed/
+# Note: requests in a sandboxed iframe are considered cross origin requests, for which, in addition to integrity checks,
+# also cors headers must be checked, see https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity.
```
What CORS headers must be sent with what values?

danimoh replied:
https://github.com/nimiq/keyguard/pull/524#discussion_r2182339196

sisou commented on src/request/sign-multisig-transaction/SignMultisigTransaction.js:
```diff
@@ -263,6 +263,10 @@ class SignMultisigTransaction {
             }
         }
+        if (ownCommitmentPairs.length !== ownSigner.commitments.length) {
+            reject(new Errors.InvalidRequestError('The number of secrets does not match the number of my commitments'));
```
Reword: 'The number of secrets does not match the number of this signer\'s commitments'

sisou commented on src/request/sign-multisig-transaction/SignMultisigTransactionApi.js:
```diff
@@ -141,10 +141,11 @@ class SignMultisigTransactionApi extends TopLevelApi {
-            if (!Array.isArray(object.secrets.encrypted) || object.secrets.encrypted.length < 2) {
+            if (!Array.isArray(object.secrets.encrypted)) {
+                // No need to check object.secrets.encrypted.length here, as the number will later be checked against
+                // the number of my provided commitments, after we identify them in SignMultisigTransaction.
```
Reword: the number of this signer's provided commitments

sisou commented on src/request/sign-multisig-transaction/SignMultisigTransactionApi.js:
```diff
@@ -141,10 +141,11 @@ class SignMultisigTransactionApi extends TopLevelApi {
-            // Not checking fixed length here, to stay flexible for future increases of the number of commitments
-            if (!Array.isArray(object.secrets.encrypted) || object.secrets.encrypted.length < 2) {
+            if (!Array.isArray(object.secrets.encrypted)) {
+                // No need to check object.secrets.encrypted.length here, as the number will later be checked against
```
Although failing here would fail in the beginning of the request, instead of _after_ password entry.

danimoh replied:
See https://github.com/nimiq/keyguard/pull/458#discussion_r2181074120

sisou commented on src/lib/Key.js:
```diff
@@ -30,11 +30,11 @@ class Key {
      * @returns {boolean}
      */
-    static _areEncryptionKeyParamsEqual(paramsA, paramsB) {
+    static _areEqualRsaKeyParams(paramsA, paramsB) {
```
Suggestion for reword: `_areRsaKeyParamsEqual`

sisou commented on src/lib/Key.js:
```diff
@@ -43,11 +43,11 @@ class Key {
      */
-    static _areDefaultEncryptionKeyParams(params) {
-        return Key._areEncryptionKeyParamsEqual(params, Key.defaultEncryptionKeyParams);
+    static _areDefaultRsaKeyParams(params) {
```
Suggestion for reword: `_areRsaKeyParamsDefault`

sisou committed:
Rework transaction data formatting

- Allow all transactions to basic accounts to interpret the recipient data as UTF8
- But join it with a contract withdrawal prefix when relevant
- Add a separate translations for adding stake that is not "Staking creation"

sisou committed:
Minor rewordings

sisou committed:
Verify RSA-key-creating forge.min.js file is original

sisou commented on client/src/PublicRequest.ts:
```diff
@@ -530,9 +530,9 @@ export type DerivePolygonAddressResult = {
-    permissions: KeyguardCommand[],
     requestedKeyPaths: string[],
     challenge: string,
+    // permissions: KeyguardCommand[],
```
Permissions are stored and verified in the Hub.

sisou committed:
Move filtering of other signers up

To make it clearer why object-comparison works.

sisou committed:
Re-add guard against same-frame messages

Which are sometimes triggered by browser extensions.

sisou committed:
Remove alias to `SignatureResult` result type

sisou committed:
Add tests for RequestParser.parseLogoUrl's new `allowEmpty` option

sisou committed:
Exclude sandboxed RSA iframe code from test includes

It was throwing an error that it must be run in a sandboxed iframe at the end of the test run.

sisou committed:
Fix and re-enable Key partial signature test

sisou committed:
Fix SignMultisigTransaction demo

sisou committed:
Fix RSAKeys demo

sisou committed:
Fix ConnectAccount demo

sisou committed:
Tabs => 4 spaces in demo files

sisou committed:
Use <script type="module"> instead of using the onload event in demo files

sisou committed:
Clean up TransactionDataFormatting

sisou committed:
Merge pull request #532 from nimiq/soeren/multisig-fixes-fixes

Multisig-fixes Fixes
